patches/api/0421-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,2335 @@
++    }
++
++    /**
++     * Gets the block position for
++     * the {@link org.bukkit.event.entity.EntityDamageByBlockEvent EntityDamageByBlockEvent}.
++     */
++    @NotNull
++    public BlockPosition blockPosition() {
++        return this.blockPosition;
++    }
++
++    /**
++     * Gets the direct source entity for
++     * the {@link org.bukkit.event.entity.EntityDamageByEntityEvent EntityDamageByEntityEvent}
++     */
++    @NotNull
++    public Entity directSource() {
++        return this.directSource;
++    }
++
++    /**
++     * Sets the block position tied for the {@link org.bukkit.event.entity.EntityDamageByBlockEvent EntityDamageByBlockEvent}.
++     * The block will be resolved at runtime where the event is called
++     *
++     * @param blockPosition the block position
++     * @return a new event context
++     */
++    @NotNull
++    public static EventContext fromBlock(@NotNull Position blockPosition) {
++        return new EventContext(blockPosition.toBlock(), null);
++    }
++
++    /**
++     * Sets the entity tied for the {@link org.bukkit.event.entity.EntityDamageByEntityEvent EntityDamageByEntityEvent}.
++     * The entity is needed to support fully the bukkit event notably the fact that some
++     * entity for example a lightning bolt doesn't have an entity source in their damage
++     * origin but is still caught by Bukkit
++     *
++     * @param directSource the direct source entity
++     * @return a new event context
++     */
++    @NotNull
++    public static EventContext fromEntity(@NotNull Entity directSource) {
++        return new EventContext(null, directSource);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/entity/damageorigin/type/DamageEffect.java b/src/main/java/io/papermc/paper/entity/damageorigin/type/DamageEffect.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..4efa33afdd16dd8e3c1ebbbdce2bf3e44d1d7f1e
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/damageorigin/type/DamageEffect.java
+@@ -0,0 +1,35 @@
++package io.papermc.paper.entity.damageorigin.type;
++
++import org.bukkit.Sound;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Damage effect is used to determine the hurt sound played
++ * by this type of damage when a played is damaged
++ */
++public enum DamageEffect {
++
++    HURT(Sound.ENTITY_PLAYER_HURT),
++    THORNS(Sound.ENCHANT_THORNS_HIT),
++    DROWNING(Sound.ENTITY_PLAYER_HURT_DROWN),
++    BURNING(Sound.ENTITY_PLAYER_HURT_ON_FIRE),
++    POKING(Sound.ENTITY_PLAYER_HURT_SWEET_BERRY_BUSH),
++    FREEZING(Sound.ENTITY_PLAYER_HURT_FREEZE);
++
++    private final Sound sound;
++
++    DamageEffect(Sound sound) {
++        this.sound = sound;
++    }
++
++    /**
++     * Gets the produced sound when a damage origin
++     * holding this type will be applied to another
++     * player
++     * @return the sound produced
++     */
++    @NotNull
++    public Sound getHurtSound() {
++        return this.sound;
++    }
++}
```

Comment:
Despite this being defined as an enum in MC, I feel like we should instead make a sort of "fake" registry that this is backed by instead.

And instead of manually defining these sounds, it should use the actual backing sound used in NMS.

Reply:
What do you mean by fake registry ? ideally i would wait for the bukkit sound to have a real registry and not a fake enum one to simplify the conversion. Actually the sound are tested against the vanilla one so they shouldn't have desync between them

Reply:
Like, currently in the NMS type they have an identifier and a sound. I feel like making this not an enum and something similar to that would be good as to handle if this is made to be data-driven in the future.